Create new validation pattern

StephenGill opened the pull request:
This PR creates the new validation pattern and makes minor changes to two linked pages

joelanman commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,90 @@
+
++ information that can’t be correct
++ information that’s too ambiguous for you to use
++ missing information, it the information is required
```
typo - 'it'

36degrees commented:
The tests are failing because there are 8 broken links:

```
patterns/validation/index.html >>> href: "/../patterns/problem-with-the-service-pages/", text: "‘there is a problem with the service’ pages"
patterns/validation/index.html >>> href: "/../patterns/page-not-found-pages/", text: "‘page not found’ pages"
patterns/validation/index.html >>> href: "/../patterns/service-unavailable-pages/", text: "‘service unavailable’ pages"
patterns/validation/index.html >>> href: "../components/error-summary/", text: "error summary"
patterns/validation/index.html >>> href: "../components/error-messages/", text: "error messages"
patterns/validation/index.html >>> href: "../error-message#be-clear-and-concise", text: "writing good error messages"
components/error-summary/index.html >>> href: "/../patterns/validation/", text: "validation pattern"
components/error-message/index.html >>> href: "/../patterns/validation/", text: "validation pattern"
```

36degrees commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,90 @@
++ show them the page again, with the form fields as the user filled them in
++ add ‘Error: ’ to the beginning of the page `<title>` so screen readers read it out as soon as possible
++ show an [error summary](/components/error-summary) at the top of the page
++ show [error messages](/components/error-messages) next to fields with errors
```
This link is broken – it should be:

```suggestion
+ show [error messages](/components/error-message/) next to fields with errors
```

joelanman commented on src/patterns/validation/index.md.njk:
```diff
@@ -8,83 +8,87 @@ backlog_issue_id:
+To start with, avoid validating the information in a field while the user is still typing it in. This sort of validation can cause problems - especially for users who type more slowly.
 
-Show the error messages when they click the ‘continue’ or ‘submit’ button at the bottom of the page. Unless your user research tells you there’s a good reason to do so, avoid showing the error message:
+Only add this sort of validation if your user research shows that, on balance, it solves more problems for users than it causes. For example, the [character count component](/components/character-count/) shows users an error message when they go over the character limit. Because it’s important that users do not spend time and effort writing out a response that turns out to be too long.
```
this is great, nice work

StephenGill commented:
> I'd also challenge the use of "typing" over "entering" the information. I don't think this guidance should just apply to keyboard entry. There are a range of other interfaces this could also apply to, for example sliders or calendars.

Thanks - I've changed it to say 'don't validate while the user is still entering information'

StephenGill commented:
> The guidance for this component is an excellent example of how to implement validation during typing, what to consider and when not to do it. It might be worth either saying that here or surfacing some of that content instead.

Agree - though on the new validation page, I think it's more about giving a high level example of a legitimate need for validation while you type, rather than details of how to implement it. And I think we've done that ok?

StephenGill commented:
> Being a bit picky but not sure what "To start with" adds here.
> 
> I also think we might need another heading, possibly a sub-heading below "When to tell the user about validation errors" for validating during input.

It's a matter of opinion, but I think the subheads look ok as they are

StephenGill commented:
Think this is ready for merging - @joelanman would you mind giving it a final review?

fofr commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,95 @@
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
++ do not add ‘required’ to your input fields
+
+## Help improve this page
```
From the preview it looks like there's two "Help improve this page" sections:
https://deploy-preview-1164--govuk-design-system-preview.netlify.com/patterns/validation/

![Screen Shot 2020-03-04 at 09 18 31](https://user-images.githubusercontent.com/319055/75863919-2360ae80-5df9-11ea-8c2d-f0af8239fc85.png)

StephenGill replied:
Thanks - I forgot that it adds a 'Help improve this page' paragraph automatically. Should be fixed now.

paulwaitehomeoffice commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
++ do not add ‘required’ to your input fields
```
Do users have to refrain from using the `required` attribute? `novalidate` disables in-browser validation triggered by it, as far as I can see, and I'm fairly sure some screen readers will say "required" when reading form fields with this attribute, which might (or, admittedly, might not) be useful for screen reader users.

joelanman replied:
Thanks for this, we'll recommend not using `required` for now as that's consistent with out approach in general (to only highlight optional fields). However we've added a line to the Research section - to check whether this is the right approach for screen reader users.

edwardhorsford commented on src/components/error-message/index.md.njk:
```diff
@@ -1,6 +1,6 @@
 ---
 title: Error message
-description: Use an error message when there is a validation error. Explain what went wrong and how to fix it
+description: When there's a validation error, use an error message to explain what went wrong and how to fix it
```
```suggestion
description: When there’s a validation error, use an error message to explain what went wrong and how to fix it
```

36degrees replied:
The Markdown parser we're using takes care of quotes automatically, converting them to the appropriate character.

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+---
+title: Recover from errors (validation)
+description: Check the answers users give to make sure they’re valid - and if there’s an error, tell them what's wrong and how to fix it
```
```suggestion
description: Check the answers users give to make sure they’re valid - and if there’s an error, tell them what’s wrong and how to fix it
```

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+layout: layout-pane.njk
+---
+
+Check the information the user gives you to make sure it’s valid. If there's a validation error, tell the user what's gone wrong and how to fix it. Turn off HTML5 validation.
```
```suggestion
Check the information the user gives you to make sure it’s valid. If there’s a validation error, tell the user what's gone wrong and how to fix it. Turn off HTML5 validation.
```

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+
+## When to use this pattern
+
+Validation should be a last resort. Minimise your chances of needing to show an error message by:
```
> Validation should always be a last resort

I find this odd phrasing. Services should try to do the hard work so it's less common - but validation is expected on the majority of service pages. Of the services I previously assessed, validation was often forgotten about / not given enough attention.

joelanman replied:
Thank yes, we've updated the language here.

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+Validation should be a last resort. Minimise your chances of needing to show an error message by:
+
++ [making sure your questions are well designed](https://www.gov.uk/service-manual/design/designing-good-questions)
++ accepting information in different formats, as long as it’s not ambiguous (for example, you should accept postcodes with or without spaces - or names with non-alphabetical characters like accents)
```
> alphabetical 

I think these are diacritics. Aren't letters like èáö still alphabetical?

edwardhorsford replied:
Could include some examples of diacritic characters.

StephenGill replied:
Thanks @edwardhorsford  - have clarified that validation should accept diacritics as well as non-alphabetical characters

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+
+### How to tell the user about validation errors
+
+If the user's answers fail validation:
```
```suggestion
If the user’s answers fail validation:
```

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
```
```suggestion
+ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your `form` tags
```

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+To turn off HTML5 validation:
+
++ add ‘[novalidate](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-novalidate)’ to your form tags
++ do not add ‘required’ to your input fields
```
I think this is unrelated to HTML5 validation.

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+Avoid adding client side validation unless you can establish that there’s a need for it. For example, because user research shows it’s helpful to users if you validate the information they’re providing as they type.
+
+If you do use client side validation, make sure your client side and server side validation rules do not conflict.
+
```
Is it worth mentioning some of the common pitfalls of client side validation?

For instance:
* Distracting / jarring / trigger to easily
* Reflows the page

joelanman replied:
We don't want to get too much into the technical aspects of client side validation, as we're not recommending it. If teams find a need to add it, they will have to look into accessibility, reliability etc for themselves. If common user needs are found, we could document tried and tested approaches.

edwardhorsford commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,92 @@
+## When to use this pattern
+
+Validation should be a last resort. Minimise your chances of needing to show an error message by:
+
```
Could add something about reasonably normalising input, for example by:
* Stripping leading and trailing whitespace
* Converting / stripping characters you understand but don't need - eg brackets in telephone numbers.
* Converting case on things like postcode and reference numbers.

joelanman replied:
Normalising might be something to cover in other guidance - for now we just want to focus on validation. Another related area we might get into later is sanitation - making input safe to store and display, but yes we wanted to keep the scope just on validation here.

timpaul commented on src/patterns/validation/index.md.njk:
```diff
@@ -0,0 +1,93 @@
+---
+title: Recover from errors (validation)
```
Could we tweak the title to remove the parentheses please? Ta

edwardhorsford commented on src/components/error-summary/index.md.njk:
```diff
@@ -21,7 +21,14 @@ Always show an error summary when there is a validation error, even if there’s
++ move keyboard focus to the error summary
++ include the heading ‘There is a problem’
++ link to each of the answers that have validation errors
++ make sure the error messages in the error summary are worded the same as those which appear next to the inputs with errors
```
In most cases I reckon this is correct - but we had cases when working on Passports where it made sense for the error next to the field to be different from the error in the summary list. I'll try to find an example.

StephenGill commented:
Should be ready for merging now. I've:

- removed the parentheses from the h1
- clarified that name validation should accept diacritics as well an non-alphabetical characters, as per, @edwardhorsford's suggestion
- added a couple of points back into the checklist about how and when to display the error summary component
- made some minor tweaks for clarity

@joelanman @timpaul

joelanman commented:
Thanks @StephenGill, looks good to me

36degrees commented:
Can we rebase this and try and squash down from 20 commits before merging please.
